cmd/joy/release_promote_test.go

```diff
@@ -0,0 +1,205 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"slices"
+	"testing"
+
+	"github.com/nestoca/joy/internal/git"
+
+	"github.com/acarl005/stripansi"
+	"github.com/spf13/cobra"
+
+	"github.com/nestoca/joy/internal/config"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/nestoca/joy/internal/testutils"
+	"github.com/nestoca/joy/pkg/catalog"
+)
+
+func executeReleasePromoteCommand(t *testing.T, cmd *cobra.Command, args ...string) (string, string, error) {
+	dir := testutils.CloneToTempDir(t, "joy-release-promote-test")
+
+	cat, err := catalog.Load(dir, nil)
+	assert.NoError(t, err)
+
+	cfg, err := config.Load(dir, dir)
+	assert.NoError(t, err)
+
+	ctx := config.ToContext(catalog.ToContext(context.Background(), cat), cfg)
+
+	var buffer bytes.Buffer
+	cmd.SetOut(&buffer)
+	cmd.SetArgs(args)
+	cmd.SetContext(ctx)
+
+	err = cmd.Execute()
+	return dir, stripansi.Strip(buffer.String()), err
+}
+
+func TestPromoteAllReleasesFromStagingToProd(t *testing.T) {
+	dir, output, err := executeReleasePromoteCommand(t, NewReleasePromoteCmd(PromoteParams{PreRunConfigs: make(PreRunConfigs)}),
+		"--source", "staging",
+		"--target", "prod",
+		"--no-prompt",
+		"--draft",
+		"--template-var", "triggeredBy=john.doe",
+		"foo")
+	fmt.Println(output)
+	require.NoError(t, err)
+
+	prURLRegex := regexp.MustCompile(`Created draft pull request: https://github.com/nestoca/joy-release-promote-test/pull/([0-9]+)`)
+	match := prURLRegex.FindStringSubmatch(output)
+	require.NotNil(t, match)
+	prURL := match[1]
+	t.Cleanup(func() {
+		closePR(t, dir, prURL)
```

Comment:
I think in this case, you don't need to capture the variables in the closure. 

This should do the trick:
```go
defer closePR(t, dir, prURL)
```

Reply:
What's the difference between `t.Cleanup()` and `defer`, any clue why they introduced `t.Cleanup()`?

Reply:
So I had to read about this. `t.Cleanup` is needed for tests marked with t.Parrallel(), because the testing framework needs to be aware of the subtests. 

The function passes through all of the subtests and registers them to run, if all subtests are marked as parrallel they will continue async and the main function will exit before the subtests finish and so parent defers would run before their subtests finish potentially.

Hence, t.Cleanup is to be used for scheduling cleanup when you have parrallel subtests.

It's quite `niché`